File: core/src/com/broll/gainea/server/core/objects/buffs/Buff.java
package com.broll.gainea.server.core.objects.buffs;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public abstract class Buff<T extends BuffableValue> {

    private Set<T> buffedValues = new HashSet<>();

    void register(T buffableValue) {
        this.buffedValues.add(buffableValue);
    }

    public List<Object> getAffectedObjects() {
        return buffedValues.stream().map(BuffableValue::getObject).distinct().collect(Collectors.toList());
    }

    public void remove() {
        this.buffedValues.forEach(it -> it.clearBuff(this));
    }

}

File: server/src/main/com/broll/gainea/server/core/objects/buffs/Buff.kt
package com.broll.gainea.server.core.objects.buffs

import com.broll.gainea.server.core.Game


abstract class Buff<V> {
    var turnsActive = 0

    private val buffedValues = mutableSetOf<BuffableValue<*, V, Buff<V>>>()
    fun register(buffableValue: BuffableValue<*, V, Buff<V>>) {
        buffedValues.add(buffableValue)
    }

    val affectedObjects: List<Any>
        get() = buffedValues.map { it.target!! }


    fun remove() = buffedValues.forEach { it.clearBuff(this) }

}

fun Buff<*>.roundsActive(game: Game) = turnsActive / game.allPlayers.size